add `[%exclave]` support (#51)
ocaml-flambda/ocaml-jst: jane/doc/local-reference.md

upon the function's return, we can rewrite the example as follows:

```
let f (local_ x) =
  let local_ y = (complex computation on x) in
  if y then [%exclave] (local_ None)
  else [%exclave] (local_ (Some x))
```

The new primitive `[%exclave]` terminates the current region early and executes
the subsequent code in the outer region. In this example, the function `f`
retains its own region where the allocation for the complex computation occurs.
This region is terminated by `[%exclave]`, releasing all temporary allocations.
Both `local_ None` and `local_ (Some x)` are considered "local" relative to the
outer region and are allowed to escape. In summary, we have temporary
allocations on the stack that are promptly released and result allocations on
the stack that can escape.

Here is another example in which the stack usage can be improved asymptotically
by applying `[%exclave]`:

```
let rec maybe_length p l = local_
  match l with
  | [] -> Ok { res = 0 }
  | x :: xs ->
      match p x with
      | false -> None
      | true ->
          match count_result f xs with
          | None as r -> r
          | Some { res = count } ->
              Some { res = count + 1 }

> ```

Reviewer: Let's just take this example out for now. I'll add a better one in a later PR.

Author: is it better to have a silly but working example, than to have no example?

The function is intended to have the type:

```
val maybe_length : ('a -> bool) -> 'a list -> local_ int option
```

This function computes the length of the list. The predicate can return `false`,
in which case the result of the entire function would be `None`. It is designed
not to allocate heap memory, instead using the stack for all `Some` allocations.
However, it will currently use O(N) stack space because all allocations occur in
the original caller's stack frame. To improve its space usage, we remove the
`local_` annotation (so the function has its own region), and wrap `Some {res =
count + 1}` inside `[%exclave]` to release the region before the allocation.
After the revision, the function should use O(1) stack space.

> `[%exclave]` terminates the current region, so values within it cannot be used

Reviewer: ```suggestion
`[%exclave]` terminates the current region, so local values from that region cannot be used
```
